### PR TITLE
Goalie Strafe

### PR DESCRIPTION
**Pull request by SamMcFarlane**

### CODE REVIEW

**Comment by NDxiura** [comment]
This needs to go in a separate file, at the moment this overwrites the parameters used in the walk

File: module/motion/QuinticWalk/src/QuinticWalk.cpp [hunk 1]
```diff
@@ -148,16 +184,16 @@ namespace module::motion {
 
             // the engine expects orders in [m] not [m/s]. We have to compute by dividing by step frequency which is
             // a double step factor 2 since the order distance is only for a single step, not double step
-            const float factor             = (1.0 / (params.freq)) / 2.0;
+            const float factor             = (1.0 / (current_config.params.freq)) / 2.0;
```
Comment (KipHamiltons) [inline module/motion/QuinticWalk/src/QuinticWalk.cpp:187]:
```suggestion
            const float factor             = (1.0f / (current_config.params.freq)) * 0.5f;
```

File: module/motion/QuinticWalk/src/QuinticWalk.cpp [hunk 2]
```diff
@@ -169,10 +205,10 @@ namespace module::motion {
                                 command.y(),
                                 command.z(),
                                 command.x() + command.y(),
-                                config.max_step[0] / factor,
-                                config.max_step[1] / factor,
-                                config.max_step[2] / factor,
-                                config.max_step_xy / factor));
+                                current_config.max_step[0] / factor,
+                                current_config.max_step[1] / factor,
+                                current_config.max_step[2] / factor,
```
Comment (KipHamiltons) [inline module/motion/QuinticWalk/src/QuinticWalk.cpp:210]:
```suggestion
                                current_config.max_step.x() / factor,
                                current_config.max_step.y() / factor,
                                current_config.max_step.z() / factor,
```

File: module/motion/QuinticWalk/src/QuinticWalk.cpp [hunk 3]
```diff
@@ -32,96 +34,130 @@ namespace module::motion {
...
+        config.params.trunk_pitch_p_coef_turn = 1.0f + cfg["walk"]["trunk"]["pitch_p_coef"]["turn"].as<Expression>();
+        config.params.kick_length             = cfg["walk"]["kick"]["length"].as<float>();
+        config.params.kick_phase              = cfg["walk"]["kick"]["phase"].as<float>();
+        config.params.kick_vel                = cfg["walk"]["kick"]["vel"].as<float>();
+        config.params.pause_duration          = cfg["walk"]["pause"]["duration"].as<float>();
+
+
+        config.max_step[0] = cfg["max_step"]["x"].as<float>();
+        config.max_step[1] = cfg["max_step"]["y"].as<float>();
+        config.max_step[2] = cfg["max_step"]["z"].as<float>();
```
Comment (KipHamiltons) [inline module/motion/QuinticWalk/src/QuinticWalk.cpp:73]:
```suggestion
        config.max_step.x() = cfg["max_step"]["x"].as<float>();
        config.max_step.y() = cfg["max_step"]["y"].as<float>();
        config.max_step.z() = cfg["max_step"]["z"].as<float>();
```

File: module/motion/QuinticWalk/src/QuinticWalk.hpp [hunk 4]
```diff
@@ -36,14 +38,24 @@ namespace module::motion {
...
             bool imu_active;
             float imu_pitch_threshold;
             float imu_roll_threshold;
-        } config;
+
+            WalkingParameter params{};
+
+            std::map<utility::input::ServoID, float> jointGains{};
+            std::vector<std::pair<utility::input::ServoID, float>> arm_positions{};
+        } normal_config, goalie_config;
```
Comment (KipHamiltons) [inline module/motion/QuinticWalk/src/QuinticWalk.hpp:53]:
```suggestion
        } normal_config{}, goalie_config{};
```

File: module/motion/QuinticWalk/src/QuinticWalk.hpp [hunk 5]
```diff
@@ -36,14 +38,24 @@ namespace module::motion {
...
             const std::vector<std::pair<utility::input::ServoID, float>>& joints);
 
-        struct {
+        struct Config {
             Eigen::Vector3f max_step;
             float max_step_xy;
 
             bool imu_active;
             float imu_pitch_threshold;
             float imu_roll_threshold;
```
Comment (KipHamiltons) [inline module/motion/QuinticWalk/src/QuinticWalk.hpp:47]:
```suggestion
            Eigen::Vector3f max_step = Eigen::Vector3f::Zero();
            float max_step_xy = 0.0f;

            bool imu_active = true;
            float imu_pitch_threshold = 0.0f;
            float imu_roll_threshold = 0.0f;
```

File: module/motion/QuinticWalk/data/config/goalie/QuinticWalk.yaml [hunk 6]
```diff
@@ -0,0 +1,129 @@
+walk:
```
[the remark this answers is not in the document]
Comment (CMurtagh-LGTM) [reply module/motion/QuinticWalk/data/config/goalie/QuinticWalk.yaml:1]: Currently yes, but this should be fixed in @ysims's behavior changes.

File: module/behaviour/planning/SimpleWalkPathPlanner/src/SimpleWalkPathPlanner.cpp [hunk 7]
```diff
@@ -254,8 +254,7 @@ namespace module::behaviour::planning {
 
 
                 std::unique_ptr<WalkCommand> command =
-                    std::make_unique<WalkCommand>(subsumptionId,
-                                                  Eigen::Vector3d(finalForwardSpeed, finalSideSpeed, angle));
+                    std::make_unique<WalkCommand>(subsumptionId, Eigen::Vector3d(forwardSpeed, sideSpeed, turnSpeed));
```
[the remark this answers is not in the document]
Comment (KipHamiltons) [reply module/behaviour/planning/SimpleWalkPathPlanner/src/SimpleWalkPathPlanner.cpp:257]: @SamMcFarlane what was the idea behind this? 
If you're keen to make this change (which I am not sure that you are), then you should remove the calculations of `finalForward/SideSpeed` above, including the other numbers used to calculate them.
Honestly, I wouldn't be against removing the commented out logs around here too ;)

[the remark this answers is not in the document]
Comment (SamMcFarlane) [reply module/behaviour/planning/SimpleWalkPathPlanner/src/SimpleWalkPathPlanner.cpp]: Those values for only for testing purposes. Accidentally committed and pushed them, will go and change them back. I'll also remove the unnecessary comments.

File: module/motion/QuinticWalk/src/QuinticWalk.cpp [hunk 8]
```diff
@@ -32,96 +34,132 @@ namespace module::motion {
...
+        config.params.trunk_x_offset_p_coef_forward = cfg["walk"]["trunk"]["x_offset_p_coef"]["forward"].as<float>();
+        config.params.trunk_x_offset_p_coef_turn    = cfg["walk"]["trunk"]["x_offset_p_coef"]["turn"].as<float>();
+        config.params.trunk_pitch_p_coef_forward =
+            1.0f + cfg["walk"]["trunk"]["pitch_p_coef"]["forward"].as<Expression>();
+        config.params.trunk_pitch_p_coef_turn = 1.0f + cfg["walk"]["trunk"]["pitch_p_coef"]["turn"].as<Expression>();
+        config.params.kick_length             = cfg["walk"]["kick"]["length"].as<float>();
+        config.params.kick_phase              = cfg["walk"]["kick"]["phase"].as<float>();
+        config.params.kick_vel                = cfg["walk"]["kick"]["vel"].as<float>();
+        config.params.pause_duration          = cfg["walk"]["pause"]["duration"].as<float>();
+
```
Comment (KipHamiltons) [inline module/motion/QuinticWalk/src/QuinticWalk.cpp:71]:
```suggestion
```

File: module/motion/QuinticWalk/src/QuinticWalk.cpp [hunk 9]
```diff
@@ -32,96 +34,132 @@ namespace module::motion {
...
+        config.max_step.y() = cfg["max_step"]["y"].as<float>();
+        config.max_step.z() = cfg["max_step"]["z"].as<float>();
+        config.max_step_xy = cfg["max_step"]["xy"].as<float>();
+
+        config.imu_active          = cfg["imu"]["active"].as<bool>();
+        config.imu_pitch_threshold = 1.0f + cfg["imu"]["pitch"]["threshold"].as<float>();
+        config.imu_roll_threshold  = cfg["imu"]["roll"]["threshold"].as<float>();
+
+        for (int i = 0; i < ServoID::NUMBER_OF_SERVOS; ++i) {
+            if ((i >= 6) && (i < 18)) {
```
Comment (KipHamiltons) [inline module/motion/QuinticWalk/src/QuinticWalk.cpp:83]: Numbers are magic. Probably name these constants. idk.

File: module/motion/QuinticWalk/src/QuinticWalk.hpp [hunk 10]
```diff
@@ -36,14 +38,24 @@ namespace module::motion {
...
+
+            std::map<utility::input::ServoID, float> jointGains{};
+            std::vector<std::pair<utility::input::ServoID, float>> arm_positions{};
+        } normal_config{}, goalie_config{};
 
-            bool imu_active;
-            float imu_pitch_threshold;
-            float imu_roll_threshold;
-        } config;
+        void load_quintic_walk(const ::extension::Configuration& cfg, Config& config) const;
```
Comment (KipHamiltons) [inline module/motion/QuinticWalk/src/QuinticWalk.hpp:55]:
```suggestion
        static void load_quintic_walk(const ::extension::Configuration& cfg, Config& config);
```

File: module/motion/QuinticWalk/src/QuinticWalk.cpp [hunk 11]
```diff
@@ -32,96 +34,132 @@ namespace module::motion {
     using utility::input::ServoID;
     using utility::motion::kinematics::calculateLegJoints;
 
+    /**
+     * @brief loads the configuration from cfg into config
+     * @param cfg A Configuration provided by the Configuration extension
+     * @param config The destination configuration that we will write to.
+     */
+    static void QuinticWalk::load_quintic_walk(const Configuration& cfg, Config& config) {
```
Comment (KipHamiltons) [inline module/motion/QuinticWalk/src/QuinticWalk.cpp:42]:
```suggestion
    void QuinticWalk::load_quintic_walk(const Configuration& cfg, Config& config) {
```

**Review by KipHamiltons — CHANGES_REQUESTED** [review]

File: module/motion/QuinticWalk/src/QuinticWalk.cpp (hunk 9 above)
Comment (SamMcFarlane) [reply module/motion/QuinticWalk/src/QuinticWalk.cpp]: Changed magic numbers to actual servos, changed variable name

Comment (KipHamiltons) [reply module/motion/QuinticWalk/src/QuinticWalk.cpp]: Great work! I'll make one more suggestion to put the cream on top

File: module/motion/QuinticWalk/src/QuinticWalk.cpp [hunk 12]
```diff
@@ -32,96 +34,131 @@ namespace module::motion {
...
+        config.imu_active          = cfg["imu"]["active"].as<bool>();
+        config.imu_pitch_threshold = 1.0f + cfg["imu"]["pitch"]["threshold"].as<float>();
+        config.imu_roll_threshold  = cfg["imu"]["roll"]["threshold"].as<float>();
+
+        for (int servo_index_number = 0; servo_index_number < ServoID::NUMBER_OF_SERVOS; ++servo_index_number) {
+            if ((servo_index_number >= ServoID::R_HIP_YAW) && (servo_index_number < ServoID::HEAD_YAW)) {
+                config.jointGains[servo_index_number] = cfg["gains"]["legs"].as<float>();
+            }
+            if (servo_index_number < ServoID::R_HIP_YAW) {
+                config.jointGains[servo_index_number] = cfg["gains"]["arms"].as<float>();
```
Comment (KipHamiltons) [inline module/motion/QuinticWalk/src/QuinticWalk.cpp:86]:
```suggestion
        for (int id = 0; id < ServoID::NUMBER_OF_SERVOS; ++id) {
            // Sets the leg gains
            if ((id >= ServoID::R_HIP_YAW) && (id < ServoID::HEAD_YAW)) {
                config.jointGains[id] = cfg["gains"]["legs"].as<float>();
            }
            // Sets the arm gains
            if (id < ServoID::R_HIP_YAW) {
                config.jointGains[id] = cfg["gains"]["arms"].as<float>();
```

**Review by KipHamiltons — APPROVED** [review]

**Pull request merged**